stubs/nvml_stubs.c

```diff
@@ -434,100 +433,79 @@ CAMLprim value
 stub_nvml_device_get_pgpu_metadata(value ml_interface, value ml_device)
 {
     CAMLparam2(ml_interface, ml_device);
-    CAMLlocal1(ml_pgpu_metadata);
+    CAMLlocal1(ml_metadata);
     nvmlReturn_t error;
     nvmlInterface *interface;
     nvmlDevice_t device;
-    unsigned int pgpuMetadataSize = 0;
-    nvmlVgpuPgpuMetadata_t *pgpuMetadata = NULL;
+    unsigned int metadataSize = 0;
+    nvmlVgpuPgpuMetadata_t *metadata = NULL;
 
     interface = (nvmlInterface *) ml_interface;
     device = *(nvmlDevice_t *) ml_device;
 
-    // Get metadata dynamically increasing the buffer size
-    int dummy;
-    do {
-        error = interface->deviceGetVgpuMetadata(device,
-                                                 (pgpuMetadata) ?
-                                                 pgpuMetadata
-                                                 :
-                                                 (nvmlVgpuPgpuMetadata_t
-                                                  *) & dummy,
-                                                 &pgpuMetadataSize);
-        if ((error == NVML_ERROR_INSUFFICIENT_SIZE)
-            && (pgpuMetadataSize > 0)) {
-            if (pgpuMetadata) {
-                free(pgpuMetadata);
-            }
-            pgpuMetadata = (nvmlVgpuPgpuMetadata_t *)
-                malloc(pgpuMetadataSize);
-            if (!pgpuMetadata) {
-                check_error(interface, NVML_ERROR_MEMORY);
-            }
-        }
+    error =
+        interface->deviceGetVgpuMetadata(device, metadata, &metadataSize);
+    if (error == NVML_SUCCESS) {
+        check_error(interface, NVML_ERROR_MEMORY);      /* should not happen */
```

Comment:
Although not possible according to the API docs, if it's a success, could this lead to a leak of the metadata if it's not freed?

Reply:
At this point no data was allocated (`metadataSize =- 0`), so I don't think this is a concern.

Reply:
I see so we expect NVML_ERROR_INSUFFICIENT_SIZE and if we don't get that something is wrong with the NVML library.